bootstrap/src/sun/nio/ch/lincheck/Types.java

```diff
@@ -0,0 +1,311 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2025 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package sun.nio.ch.lincheck;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class Types {
+    private static class Utils {
+        /**
+         * Calls `consumer` for each descriptor substring of the `methodDesc`, including the return type of the method.
+         * @param methodDesc descriptor of the method.
+         * @param consumer callback called for each descriptor type. Passed parameter is the substring which contains the next descriptor type.
+         */
+        private static void iterateThroughDescriptors(String methodDesc, Consumer<String> consumer) {
+            // Modified code for iterating over the type descriptors
+            // in the method descriptor (which look like this: "(args...)ret")
+            int currentOffset = 1;
+
+            while (methodDesc.charAt(currentOffset) != ')') {
```

Comment:
`)` can be used in the type name

Reply:
this part of code follows how parsing is done in asm sources

Reply:
That's problem of ASM.
https://pl.kotl.in/LUX1kmmJk

Reply:
Or a problem of people who use `)` in their class names : )